server/src/main/java/com/objectcomputing/checkins/services/member_skill/skillsreport/SkillsReportServicesImpl.java

```diff
@@ -0,0 +1,191 @@
+import java.util.List;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.UUID;
+
+@Singleton
+public class SkillsReportServicesImpl implements SkillsReportServices {
+
+    private final MemberSkillRepository memberSkillRepo;
+    private final MemberProfileRepository memberProfileRepo;
+    private final SkillRepository skillRepo;
+    private final HashMap<String, Integer> skillLevels;
+
+    public SkillsReportServicesImpl(MemberSkillRepository memberSkillRepo,
+                                    MemberProfileRepository memberProfileRepo,
+                                    SkillRepository skillRepo) {
+        this.memberSkillRepo = memberSkillRepo;
+        this.memberProfileRepo = memberProfileRepo;
+        this.skillRepo = skillRepo;
+
+        skillLevels = new HashMap<>();
+        skillLevels.put("interested", 0);
+        skillLevels.put("novice", 1);
+        skillLevels.put("intermediate", 2);
+        skillLevels.put("advanced", 3);
+        skillLevels.put("expert", 4);
+    }
+
+    @NotNull
+    public SkillsReportResponseDTO report(@NotNull SkillsReportRequestDTO request) {
+        SkillsReportResponseDTO response = null;
+        if (request != null) {
+            final List<SkillLevelDTO> skills = request.getSkills();
+            final Set<UUID> members = request.getMembers();
+            final Boolean inclusive = request.isInclusive();
+
+            for (SkillLevelDTO skill : skills) {
+                if (!skillRepo.existsById(skill.getId())) {
+                    throw new BadArgException(String.format("Invalid skill ID %s", skill.getId()));
+                }
+            }
+
+            if (members != null) {
+                for (UUID member : members) {
+                    if (!memberProfileRepo.existsById(member)) {
+                        throw new BadArgException(String.format("Invalid member profile ID %s", member));
+                    }
+                }
+            }
+
+            response = new SkillsReportResponseDTO();
+
+            final List<TeamMemberSkillDTO> potentialMembers = getPotentialQualifyingMembers(skills);
+            if (members == null || members.isEmpty()) {
+                if (inclusive == null || !inclusive) {
+                    response.setTeamMembers(potentialMembers);
+                } else {
+                    final List<TeamMemberSkillDTO> qualifiedMembers = getMembersSatisfyingAllSkills(potentialMembers, skills);
+                    response.setTeamMembers(qualifiedMembers);
+                }
+            } else {
+                if (inclusive == null || !inclusive) {
+                    final List<TeamMemberSkillDTO> qualifiedMembers = removeMembersNotRequested(potentialMembers, members);
+                    response.setTeamMembers(qualifiedMembers);
+                } else {
+                    final List<TeamMemberSkillDTO> membersInList = removeMembersNotRequested(potentialMembers, members);
```

Comment:
line 74 and 77 are the same...pull outside of the if/else? for that matter, if you make the check for null or empty members at the top and only removed members in that case. I think you could simplify these nested conditions. Not a big deal, just a thought.

Reply:
> line 74 and 77 are the same...pull outside of the if/else?
Good point. Thanks.